From 3b4dd734637803a20ceb7eeaa009f753f4936542 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:32:26 -0400
Subject: [iommu] vt-d: fix resource leakage on error recovery path in iommu_init_domain

Message-id: <20140905013226.30867.38508.stgit@gir.stowe>
Patchwork-id: 91277
O-Subject: [RHEL7.1 PATCH 13/28] iommu/vt-d: fix resource leakage on error recovery path in iommu_init_domain
Bugzilla: 1129808
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129808
Upstream Status: 852bdb04f81c276969d43b9e15048259d028881f

commit 852bdb04f81c276969d43b9e15048259d028881f
Author: Jiang Liu <jiang.liu@linux.intel.com>
Date:   Mon Jan 6 14:18:11 2014 +0800

    iommu/vt-d: fix resource leakage on error recovery path in iommu_init_domain

    Release allocated resources on error recovery path in function
    iommu_init_domains().

    Also improve printk messages in iommu_init_domains().

    Acked-by: Yijing Wang <wangyijing@huawei.com>
    Signed-off-by: Jiang Liu <jiang.liu@linux.intel.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/intel-iommu.c b/drivers/iommu/intel-iommu.c
index 1f56039..0979a07 100644
--- a/drivers/iommu/intel-iommu.c
+++ b/drivers/iommu/intel-iommu.c
@@ -1255,8 +1255,8 @@ static int iommu_init_domains(struct intel_iommu *iommu)
  unsigned long nlongs;
 
  ndomains = cap_ndoms(iommu->cap);
- pr_debug("IOMMU %d: Number of Domains supported <%ld>\n", iommu->seq_id,
-   ndomains);
+ pr_debug("IOMMU%d: Number of Domains supported <%ld>\n",
+   iommu->seq_id, ndomains);
  nlongs = BITS_TO_LONGS(ndomains);
 
  spin_lock_init(&iommu->lock);
@@ -1266,13 +1266,17 @@ static int iommu_init_domains(struct intel_iommu *iommu)
   */
  iommu->domain_ids = kcalloc(nlongs, sizeof(unsigned long), GFP_KERNEL);
  if (!iommu->domain_ids) {
-  printk(KERN_ERR "Allocating domain id array failed\n");
+  pr_err("IOMMU%d: allocating domain id array failed\n",
+         iommu->seq_id);
   return -ENOMEM;
  }
  iommu->domains = kcalloc(ndomains, sizeof(struct dmar_domain *),
    GFP_KERNEL);
  if (!iommu->domains) {
-  printk(KERN_ERR "Allocating domain array failed\n");
+  pr_err("IOMMU%d: allocating domain array failed\n",
+         iommu->seq_id);
+  kfree(iommu->domain_ids);
+  iommu->domain_ids = NULL;
   return -ENOMEM;
  }
 
-- 
1.7.1